File: python-sdk/captcha_background_sdk/foreground_skew.py
from __future__ import annotations

import math
from typing import Iterable, List, Tuple

Pixel = Tuple[int, int]


def _normalize_angle(angle: float) -> float:
    while angle > 45.0:
        angle -= 90.0
    while angle < -45.0:
        angle += 90.0
    return angle


def estimate_foreground_skew(
    pixels: Iterable[Pixel],
    min_pixels: int = 20,
    max_abs_angle: float = 45.0,
) -> dict:
    sampled: List[Pixel] = list(pixels)
    if len(sampled) < min_pixels:
        return {
            "angle_degrees": 0.0,
            "confidence": 0.0,
            "pixel_count": len(sampled),
            "eigen_ratio": 0.0,
        }

    xs = [float(x) for x, _ in sampled]
    ys = [float(y) for _, y in sampled]
    mean_x = sum(xs) / len(xs)
    mean_y = sum(ys) / len(ys)

    cov_xx = 0.0
    cov_xy = 0.0
    cov_yy = 0.0
    for x, y in sampled:
        dx = float(x) - mean_x
        dy = float(y) - mean_y
        cov_xx += dx * dx
        cov_xy += dx * dy
        cov_yy += dy * dy

    cov_xx /= len(sampled)
    cov_xy /= len(sampled)
    cov_yy /= len(sampled)

    trace = cov_xx + cov_yy
    delta = max(0.0, (cov_xx - cov_yy) ** 2 + 4.0 * cov_xy * cov_xy)
    root_delta = math.sqrt(delta)

    lambda1 = 0.5 * (trace + root_delta)
    lambda2 = 0.5 * (trace - root_delta)

    angle_radians = 0.5 * math.atan2(2.0 * cov_xy, cov_xx - cov_yy)
    angle_degrees = _normalize_angle(math.degrees(angle_radians))
    angle_degrees = max(-max_abs_angle, min(max_abs_angle, angle_degrees))

    confidence = 0.0
    if lambda1 > 1e-9:
        confidence = max(0.0, min(1.0, (lambda1 - lambda2) / lambda1))

    return {
        "angle_degrees": float(angle_degrees),
        "confidence": float(confidence),
        "pixel_count": len(sampled),
        "eigen_ratio": float((lambda1 - lambda2) / lambda1) if lambda1 > 1e-9 else 0.0,
    }

File: typescript-sdk/src/foreground-skew.ts
import { Pixel } from './types';

function normalizeAngle(angle: number): number {
  while (angle > 45) angle -= 90;
  while (angle < -45) angle += 90;
  return angle;
}

export function estimateForegroundSkew(pixels: Pixel[], minPixels: number = 20, maxAbsAngle: number = 45.0) {
  const sampled = pixels.slice();
  if (sampled.length < minPixels) {
    return { angle_degrees: 0, confidence: 0, pixel_count: sampled.length, eigen_ratio: 0 };
  }

  const xs = sampled.map((p) => p[0]);
  const ys = sampled.map((p) => p[1]);
  const meanX = xs.reduce((s, v) => s + v, 0) / xs.length;
  const meanY = ys.reduce((s, v) => s + v, 0) / ys.length;

  let covXX = 0;
  let covXY = 0;
  let covYY = 0;
  for (const [x, y] of sampled) {
    const dx = x - meanX;
    const dy = y - meanY;
    covXX += dx * dx;
    covXY += dx * dy;
    covYY += dy * dy;
  }
  covXX /= sampled.length;
  covXY /= sampled.length;
  covYY /= sampled.length;

  const trace = covXX + covYY;
  const delta = Math.max(0, (covXX - covYY) ** 2 + 4 * covXY * covXY);
  const rootDelta = Math.sqrt(delta);

  const lambda1 = 0.5 * (trace + rootDelta);
  const lambda2 = 0.5 * (trace - rootDelta);

  const angleRadians = 0.5 * Math.atan2(2 * covXY, covXX - covYY);
  let angleDegrees = normalizeAngle((angleRadians * 180) / Math.PI);
  angleDegrees = Math.max(-maxAbsAngle, Math.min(maxAbsAngle, angleDegrees));

  let confidence = 0;
  if (lambda1 > 1e-9) {
    confidence = Math.max(0, Math.min(1, (lambda1 - lambda2) / lambda1));
  }

  return {
    angle_degrees: Number(angleDegrees),
    confidence: Number(confidence),
    pixel_count: sampled.length,
    eigen_ratio: lambda1 > 1e-9 ? Number((lambda1 - lambda2) / lambda1) : 0,
  };
}
